varlink: Add new varlink_collect method

arthurshau opened the pull request:
varlink_collect is meant to be used when the client is willing to wait for the reply from the varlink method, much like varlink_call. However, unlike varlink_call, it allows the client to collect all "more" replies that may be sent by method before the "final" reply is enqueued. It aggregates all of these replies into a json variant array that it returns to the client.

I added a test into test-varlink - sort of just shoehorned it into the main test. LMK if you want me to flesh it out more, or if there are any issues with the implementation.

poettering commented:
Hmm, what's the usecase for this? I mean, this is an internal subsystem so far, and your new function has no users?

arthurshau commented:
We would be able to use this for bpfd #28268  - right now, the way 'bpfctl list' command is implemented, it hangs indefinitely while waiting for 'more' replies because of the event loop, and can't exit even when the final reply has been sent. 

We thought it would be good to create a common helper function to solve this case. It would ideally make it so that all use cases that want to wait for replies that 'continue' can do so in a way similar to varlink_call, where they get what they want passed back in ret_parameters.

Of course, we can always implement this functionality in bpfctl alone if you prefer, and look to extrapolate it into a new function later if we want. Lmk what you think.

DaanDeMeyer commented on src/test/test-varlink.c:
```diff
@@ -300,13 +301,24 @@ int main(int argc, char *argv[]) {
+        assert_se(varlink_connect_address(&d, sp) >= 0);
+        assert_se(varlink_set_description(d, "collect-client") >= 0);
+
+        assert_se(varlink_collect(d, "io.test.DoSomething", v, &o) >= 0);
```
We should add a new server method that uses `VARLINK_MORE` and test against that instead so we exercise the functionality of varlink_collect() properly.

DaanDeMeyer commented on src/shared/varlink.c:
```diff
@@ -1828,6 +1828,116 @@ int varlink_callb(
+int varlink_collectb(
+                Varlink *v,
+                const char *method,
+                JsonVariant **ret_parameters,...) {
```
Missing space before `...`

DaanDeMeyer commented on src/shared/varlink.c:
```diff
@@ -1828,6 +1828,116 @@ int varlink_callb(
+        if (r < 0)
+                return log_error_errno(r, "Failed to issue method call: %m");
+
+        while (v->state == VARLINK_AWAITING_REPLY_MORE) {
```
My idea was that we'd check for `VARLINK_REPLY_CONTINUES` in the callback instead. Once that flag isn't set anymore we'd call `sd_event_exit()`. Then we don't have to rely on the internal state machine.

DaanDeMeyer commented on src/shared/varlink.c:
```diff
@@ -1828,6 +1828,116 @@ int varlink_callb(
+        if (r < 0)
+                return varlink_log_errno(v, r, "Failed to sanitize parameters: %m");
+
+        r = sd_event_default(&event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get event loop: %m");
+
+        r = sd_event_set_signal_exit(event, true);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enable exit on SIGINT/SIGTERM: %m");
```
We can't use these in library code. We should always allocate a new event loop here and not register SIGINT/SIGTERM handlers.

DaanDeMeyer commented on src/shared/varlink.c:
```diff
@@ -1828,6 +1828,116 @@ int varlink_callb(
+                JsonVariant *parameters,
+                JsonVariant **ret_parameters) {
+
+        JsonVariant *array = NULL;
```
This needs to have `_cleanup_(json_variant_unrefp)` or it will leak memory on failure

DaanDeMeyer commented on src/shared/varlink.c:
```diff
@@ -1828,6 +1828,116 @@ int varlink_callb(
+        int r;
+
+        assert_return(v, -EINVAL);
+        assert_return(method, -EINVAL);
```
This is missing an assertion that `ret_parameters` is set. That can also be renamed to `ret` as its the only output argument.

DaanDeMeyer commented on src/shared/varlink.c:
```diff
@@ -1828,6 +1828,116 @@ int varlink_callb(
+
+        r = sd_event_default(&event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get event loop: %m");
```
We should always use `varlink_log()` or `varlink_log_errno()` here and elsewhere in this file.

poettering commented on src/shared/varlink.c:
```diff
@@ -1828,6 +1828,116 @@ int varlink_callb(
+        if (r < 0)
+                return varlink_log_errno(v, r, "Failed to sanitize parameters: %m");
+
+        r = sd_event_default(&event);
```
please don't instantiate an event loop here. because a varlink object can only be associated with one event loop, and this should not already be consumed internally.

the varlink connection doesn't need the event loop really, you can just loop around varlink_process() yourself, see how varlink_call() above does it

poettering commented on src/shared/varlink.c:
```diff
@@ -1828,6 +1828,116 @@ int varlink_callb(
+                r = sd_event_run(event, UINT64_MAX);
+                if (r < 0) {
+                        return log_error_errno(r, "Failed to run event loop: %m");
+                }
```
coding style, no {} around single line if blocks

arthurshau commented:
Addressed comments, changed logic to use varlink internals rather than sd_event_loop(), and added a more fleshed out test in a separate thread in test-varlink.

arthurshau commented on src/shared/varlink.c:
```diff
@@ -1828,6 +1828,113 @@ int varlink_callb(
+                Varlink *v,
+                const char *method,
+                JsonVariant *parameters,
+                JsonVariant **ret) {
```
Is there some way I can pass all of `ret_parameters`, `ret_error_id`, and `ret_flags` to the callback through v->userdata then? That way I can return all of them to the client directly in the callback.

I was trying to do it by passing a json object made from `json_build` that would have `parameters`, `error`, and `flag` keys, but wasn't able to get it to work.

arthurshau replied:
I did it by changing userdata to point to a JSON object and utilizing json_variant_merge_object() a lot. Hope this is ok

arthurshau commented:
Added `const char **ret_error_id` and `VarlinkReplyFlags *ret_flags` parameters.

DaanDeMeyer reviewed (changes requested):
Instead of shuffelling information from the callback via a json variant, you should add a private struct `CollectContext` or so and use that to pass the return arguments to the callback via the `userdata` pointer.

arthurshau commented on src/test/test-varlink.c:
```diff
@@ -294,27 +359,40 @@ int main(int argc, char *argv[]) {
 
         assert_se(varlink_attach_event(c, e, 0) >= 0);
 
+        assert_se(pthread_create(&u, NULL, collect_thread, (void*) lp) == 0);
```
I tried putting it at the end of thread() before, but the flood test disconnects from the server, so it didn't work. But looks like shoving it in at the beginning works, so I'll do that.

arthurshau commented:
Address comments

DaanDeMeyer commented on src/shared/varlink.c:
```diff
@@ -2088,6 +2094,134 @@ int varlink_callb(
+        switch (v->state) {
+
+        case VARLINK_IDLE_CLIENT:
+                if (ret_parameters)
+                        *ret_parameters = context.parameters;
+                if (ret_error_id)
+                        *ret_error_id = context.error_id;
+                if (ret_flags)
+                        *ret_flags = context.flags;
+                return 1;
```
I would break here and move this logic to the end of the function

DaanDeMeyer reviewed (changes requested):
Almost good

arthurshau commented:
Address comments
